### DISCUSSION TITLE
Power Save vs Sleep?

### DISCUSSION THREAD

**Discussion by Fishbone69** [olikraus/u8g2#2116] [General]
Hey All.
       I am using an SH1106 and am able to get it to sleep and awaken using the u8g2.setPowerSave(1); I went back to the data sheet on the display and they discuss power save and also "sleep mode".  Are power save and sleep mode the same thing?  I am trying to use the lowest power consumption I can.  So if sleep mode is less consumption than power save, I may want to look into that.  

Thanks, smart ppl!

Fish

**Comment by olikraus** [comment]
The u8g2 power save cmd will usually only turn off the display. There are some rare cases where people contributed proper low power cmd sequence. In case of the SH1106 (which is compatible to the SSD1306) just the display is turned off: 
https://github.com/olikraus/u8g2/blob/b92f9e5464322e35154ea53fb41d2254d0256bcc/csrc/u8x8_d_ssd1306_128x64_noname.c#L215-L219

Things to consider:
 - There are multiple ways to shutdown  the display. There is a sleep mode (0x0ad cmd), but also a proper power down sequence including shutdown of the charge pump.
 - There is a trade off between time and power consumption. Power up will take longer of the charge pump had been stopped.
 - I could imagine, that full power down (display off, charge pump off) will be better then just sleep mode, but it probably needs to be measured.

**Reply by Fishbone69** [reply]
Id like to explore a full power down to include  display and charge pump off.  Is this something a novice could do?  

Thanks, Oli!

**Reply by olikraus** [reply]
Sure, why not. use the sendF command to send the shut down sequence to the display

**Reply by Fishbone69** [reply]
i'll look into it.

**Reply by Fishbone69** [reply]
Hey oli...one other thing.  I just picked up an SPI version of the SH1106 (I was using an I2C) .  I want to see what the performance differences is.  I am having trouble getting it to work.  First the pins are labeled:
Vcc
GND 
CLK
MOSI
RES
SDC
CCS

I am assuming CLK = SCL and CCS = CS.  My controller is an esp32...I went to the manufacturer's pinout here:https://www.wemos.cc/en/latest/c3/c3_pico.html
but when I ran this:
  Serial.println(MOSI);
  Serial.println(MISO);
  Serial.println(SCK);
  Serial.println(SS);  

I got different pins.  I thought you can define custom pins (via the constructor) for your board or do I have to use the defaults?  Looking at theconstructor :
U8G2_SH1106_128X64_NONAME_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
which looks like it is only looking for (rotation, cs, data, reset)?  

Any help is appreciated.

**Reply by olikraus** [reply]
First step should be the 4W_SW constructor, where you can define the pins. Once this works, go ahead with the HW constructor.

**Reply by Fishbone69** [reply]
No 4W_SW constructors are defined for the SH1106.  This is o.k. since I can move some pins around.  I tried them in both the manufacturers pinout as well as the pins obtained from the Serial.println statements above.  I did define the cs pin per the constructor.  So that means the SDC pin is dc.

**Reply by Fishbone69** [reply]
meh...still nothing.  This means I'm doing something wrong.  Could it still be looking on the I2C...I do have an I2C RTC running at the same time.

**Reply by Fishbone69** [reply]
o.k...I did find this one U8G2_SH1106_128X64_NONAME_F_4W_SW_SPI(rotation, clock, data, cs, dc [, reset]) [full framebuffer, size = 1024 bytes]

and in my code it looks like
U8G2_SH1106_128X64_NONAME_F_4W_SW_SPI u8g2(U8G2_R0, 2, 4, 7, 0);
still nothing.  
CLK on pin 2
MOSI on pin 4
CCS on pin 7
SDC on pin 0
(no reset).

Maybe it is a bad display

**Reply by olikraus** [reply]
you should connect the reset pin to an empty GPIO port and provide the Arduino number of that GPIO to the constructor as last argument.

**Reply by Fishbone69** [reply]
That seemed to have worked...BUT I got a very unexpected result.  The FPS is about HALF of what I was getting on I2C.  The I2C I was using HW vs SW but was running Full screen buffer mode on both.

**Reply by Fishbone69** [reply]
also, when putting the board to sleep I was using u8g2.setPowerSave(1);.  Upon waking the screen doesn't come back on, even with u8g2.setPowerSave(0);

**Reply by Fishbone69** [reply]
lol...o.k. so when I go to HW with a reset pin, the performance SCREAMS
U8G2_SH1106_128X64_NONAME_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 7, /* dc=*/ 0, /* reset=*/ 1);
I was getting 31 FPS with I2C.  With SPI in SW I was getting 18 FPS.  With SPI in HW I am getting 197 FPS!
However, I have used up all the gpio on my small board - it is a pico.  I can recover 1 pin if I go with an SPI RTC and put it on the same bus.  Also, any ideas on why it won't wake-up after powersave?

**Reply by Fishbone69** [reply]
o.k....I figured it out.  The pin used for the clock on the SPI was previously identified as one of the wake pins.   Everything works great now other than using up way more gpio pins.  Thanks for all the help Oli.  Do you have a place to donate to u8g2?

Fish

**Reply by olikraus** [reply]
I am happy that everything works 😃

**Comment by budger-collab** [comment]
Great question - this is exactly the kind of detail that matters when you’re optimizing for ultra-low power. Datasheets can be tricky, especially when “power save” and “sleep mode” sound similar but behave differently under the hood. Digging into how each mode handles internal clocks and refresh cycles is the right move. This mindset carries over to health too - people often ask [how does inspire work for sleep apnea](https://www.inspiresleep.com/en-us/how-inspire-therapy-works/) when they want a deeper, more efficient solution rather than just surface-level fixes.
